### ISSUE TITLE
[Feature Request] Reservoir Restart & Batch Normalisation Before Flattening

### ISSUE THREAD

**Issue by danieltudosiu** [lucidrains/vector-quantize-pytorch#11]
There are two interesting features (low implementation overhead) from the paper [Robust Training of Vector Quantized Bottleneck Models](https://arxiv.org/pdf/2005.08520.pdf):

- 3.A Importance of proper scaling - Batch normalisation 
    -  This is in line with the [Orthogonal Regularisation](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L348), but I am not sure of its interaction with it. Furthermore it will need DDP handling by using [SyncBatchNorm](https://pytorch.org/docs/stable/generated/torch.nn.SyncBatchNorm.html) more specifically the SyncBatchNorm.convert_sync_batchnorm() method
- 3.B Batch data-dependent codebook updates 
    - This is an extended logic improvement of the K-mean initialization and of the Code Restarts.    
    - Useful code is [here](https://github.com/distsup/DistSup/blob/master/distsup/modules/bottlenecks.py#L134)

**Comment by lucidrains** [comment]
hmm, to be honest, not a big fan of adding batch-dependent logic
